File: 2020/python/day1.py
with open('../day_1.txt', 'r') as f:
    expenses = list(int(s.rstrip()) for s in f)

target = 2020
lookup = set(expenses)
p1 = next(n*(target-n) for n in expenses if target-n in lookup)
p2 = next(n*m*(target-n-m) for i,n in enumerate(expenses) for m in expenses[i+1:] if target-n-m in lookup)

print('part 1:', p1)
print('part 2:', p2)

File: 2020/kotlin/day1.kt
package aoc2020

import java.io.File

fun helper(input: List<Int>, target: Int): Int {
	var b = 0
	var e = input.size-1

	while (b < e) {
		val s = input[b] + input[e]
		when {
			s == target -> return input[b] * input[e]
			s < target -> b += 1
			else -> e -= 1
		}
	}
	return -1
}

fun part1(input: List<Int>, target: Int): Int {
	return helper(input, target)
}

fun part2(input: List<Int>, target: Int): Int {
	for (i in 0..input.size-1) {
		val n = input[i]
		val m = helper(input.subList(i+1, input.size), target - n)
		if (m > 0) return n*m
	}
	return -1
}

fun main() {
	val l = File("../day_1.txt").readLines().map(String::toInt).sorted()
	println("Part 1: ${part1(l, 2020)}")
	println("Part 2: ${part2(l, 2020)}")
}
